src/main/groovy/edu/oregonstate/mist/metaxe/XEAppsResource.groovy

```diff
@@ -0,0 +1,151 @@
+@PermitAll
+@groovy.transform.TypeChecked
+class XEAppsResource extends Resource {
+    private XEAppDAO dao
+    private URI myEndpointUri
+
+    private final String JSONAPI_TYPE = "xeapp"
+    private static Pattern sanitizeRegex = ~/[^A-Za-z0-9\.\-]/
+
+    XEAppsResource(XEAppDAO dao, URI endpointUri) {
+        this.dao = dao
+        this.myEndpointUri = endpointUri
+        this.endpointUri = endpointUri
+    }
+
+    /**
+     * Get a specific application by ID.
+     *
+     * @param id the application id
+     */
+    @Timed
+    @GET
+    @Path("{id}")
+    ResultObject getById(@PathParam("id") String id) {
+        id = sanitize(id)
+
+        ESResult es = this.dao.getById(id)
+        if (es == null) {
+            throw new NotFoundException()
+        }
+
+        new ResultObject(
+            data: mapESObject(es),
+            links: [
+                self: this.urlFor(es.id)
+            ]
+        )
+    }
+
+    private ResourceObject mapESObject(ESResult es) {
+        new ResourceObject(
+            id: es.id,
+            type: JSONAPI_TYPE,
+            attributes: new Attributes(
+                applicationName: es.source.applicationName,
+                versions: es.source.versions.collectEntries { v ->
+                    [ (v.instance): v.version ]
+                }
+            ),
+            links: [
+                self: this.urlFor(es.id)
+            ],
+        )
+    }
+
+    private String urlFor(String id) {
+        UriBuilder.fromUri(this.myEndpointUri).path("xeapps/{id}").build(id).toString()
+    }
+
+    /**
+     * Get all applications, their versions, and what instances they're deployed in.
+     *
+     * @param q query string; will match all applications which contain this
+     *          string as a substring
+     * @param instance filter by deployed instance (prod, devl, dev2)
+     * @param version filter by deployed version
+     * @return a ResultObject with the list of results
+     */
+    @Timed
+    @GET
+    ResultObject search(
+        @QueryParam("q")        String q,
+        @QueryParam("instance") String instance,
+        @QueryParam("version")  String version
+    ) {
+        q = sanitize(q)
+        instance = sanitize(instance)
+        version = sanitize(version)
+
+        ESHits results = this.dao.search(q, instance, version, this.pageNumber, this.pageSize)
+
+        def params = [:]
+        if (q) {
+            params.q = q
+        }
+        if (instance) {
+            params.instance = instance
+        }
+        if (version) {
+            params.version = version
+        }
+
+        new ResultObject(
+            data: results.hits.collect { mapESObject(it) },
+            links: getPaginationLinks(params, results.total),
+        )
+    }
+
+    /**
+     * Sanitize strips everything except
+     * ascii letters, numbers, hyphen, and period.
+     */
+    private static String sanitize(String s) {
+        if (s != null) {
```

Comment:
No, i only want to return null if s is null. If s is an empty string sanitize should return an empty string.